### DISCUSSION TITLE
Unexpected error with super() and Generic types

### DISCUSSION THREAD

**Discussion by ronf** [python/typing#909] [Q&A]
I'm running into an odd problem with mypy not properly picking up a type variable when a call is made to super(). Here's a small example illustrating the problem:

```python
from typing import AnyStr, Generic

class Foo(Generic[AnyStr]):
    def method1(self, s: AnyStr) -> None:
        print(s)


class Bar(Foo[AnyStr]):
    def method1(self, s: AnyStr) -> None:
        print('Child before')
        super().method1(s)
        print('Child after')
```

When mypy is run on this, it outputs:

```
t.py:11: error: Argument 1 to "method1" of "Foo" has incompatible type "str"; expected "AnyStr"
t.py:11: error: Argument 1 to "method1" of "Foo" has incompatible type "bytes"; expected "AnyStr"
Found 2 errors in 1 file (checked 1 source file)
```

I was able to make this work by changing the definition of class Bar as follows:

```python
class Bar(Foo, Generic[AnyStr]):
```

However, I don't really understand why this is necessary or what the difference is between these two syntaxes. I've been using the first syntax here in many other places without running into any problems. Can anyone explain the difference here, and when to use one vs. the other?

**Comment by sobolevn** [comment]
This looks like a `mypy` bug to me. I will look into it 🙂

**Reply by sobolevn** [reply]
PR: https://github.com/python/mypy/pull/11378

**Reply by ronf** [reply]
Thanks very much for the quick response!

I tried applying this patch against mypy 0.910 (the latest I can find on PyPI) and I'm still seeing an error, though it is a different one than what I saw before. With the test code above, I am now getting:

```
t.py:11: error: "method1" undefined in superclass
Found 1 error in 1 file (checked 1 source file)
```

Are there other changes I need to pick up related to this issue? It looks like https://github.com/python/mypy/issues/1069 mentions a similar error but doesn't appear to be the same issue, as my parent class is not `Any`.

**Reply by sobolevn** [reply]
> I tried applying this patch against mypy 0.910 (the latest I can find on PyPI) and I'm still seeing an error, though it is a different one than what I saw before. With the test code above

Try `master` instead. It has way more new features / bugfixes than `0.910`

**Reply by ronf** [reply]
Thanks - I got a chance to move to master tonight. I ran into a few incompatibilities from 0.910 related to some definitions in hashlib.pyi and logger.pyi, but I was able to sort them out without too much trouble.

I can confirm that the PR in python/mypy#11378 eliminates the errors I was previously seeing in the example listed above.

**Reply by sobolevn** [reply]
Awesome! Happy to help! 👍
